server/src/main/java/io/crate/expression/predicate/IsNullPredicate.java

```diff
@@ -151,18 +155,30 @@ public static Query refExistsQuery(Reference ref, Context context) {
             if (fieldType != null && !fieldType.omitNorms()) {
                 return new FieldExistsQuery(field);
             }
-            if (ArrayType.unnest(ref.valueType()) instanceof ObjectType objType) {
+
+            if (ref.valueType() instanceof ObjectType objType) {
+                if (objType.innerTypes().isEmpty()) {
+                    return null;
+                }
                 BooleanQuery.Builder booleanQuery = new BooleanQuery.Builder();
                 for (var entry : objType.innerTypes().entrySet()) {
                     String childColumn = entry.getKey();
                     Reference childRef = context.getRef(ref.column().append(childColumn));
-                    Query refExistsQuery = refExistsQuery(childRef, context);
+                    Query refExistsQuery = refExistsQuery(childRef, context, true);
                     if (refExistsQuery == null) {
                         return null;
                     }
                     booleanQuery.add(refExistsQuery, Occur.SHOULD);
                 }
-                return new ConstantScoreQuery(booleanQuery.build());
+                // Even if a child columns exist, an object can have empty values, we have to run generic function.
+                // Example of such object:
+                // CREATE TABLE t (obj OBJECT as (x int));
+                // INSERT INTO t (obj) VALUES ({});
+                return new BooleanQuery.Builder()
+                    .setMinimumNumberShouldMatch(1)
+                    .add(new ConstantScoreQuery(booleanQuery.build()), Occur.SHOULD)
+                    .add(Queries.not(isNullFuncToQuery(ref, context)), Occur.SHOULD)
+                    .build();
```

Comment:
I'm not sure I follow - if there are child columns, the `refExistsQuery` on `obj.x` should figure out if the object is empty. Why does that not work?

Reply:
`obj.x` build exists query which cannot handle `{}`/empty object - `IS NULL` returns true instead of false.

Test case below fails both on master and on this PR if we remove generic and just do
`return new ConstantScoreQuery(booleanQuery.build())`
          

```
execute(" CREATE TABLE testdrive (details OBJECT as (x int))");
execute("INSERT INTO testdrive (details) VALUES ({})");
execute("INSERT INTO testdrive (details) VALUES (NULL)");
execute("REFRESH TABLE testdrive");
execute("SELECT count(*) AS count_object_is_null FROM testdrive WHERE details IS NULL");
assertThat(printedTable(response.rows()), Matchers.is("1\n"));
```

Reply:
> as far as I understand, currently we build exists query for leafs, say for obj.x which cannot handle {}/empty object - IS NULL returns true instead of false.

Okay so the problem is that in both `{}` and `NULL` the child objects are absent and therefore the query matches.

Could you compare the performance of this version with

```
      .add(new ConstantScoreQuery(booleanQuery.build()), Occur.MUST)
      .add(Queries.not(isNullFuncToQuery(ref, context)), Occur.FILTER)
```

To see one is cheaper?

Reply:
pushed an update to unit test to catch leaf case: https://github.com/crate/crate/pull/13106/commits/b64267b276ca6462b5fcd492173f48108b161fc6
to align with tricky (temporal) ITest.

Reply:
Right, I had mixed it up and thought the booleanQuery would already imply the "is null" - missed that the full thing is going to be put into a "not".

One thing that might be possible is to add the genericFunction to the `booleanQuery` instead of wrapping it in another one. Not sure if it makes a difference or if it helps with performance, so I guess keeping it as is is fine.